### ISSUE TITLE
Programmatically adding an item - Fields.push

### ISSUE THREAD

**Issue by sgentile** [erikras/redux-form#1978]
I'm wanting to show a form with a field array that by default shows the field array without clicking an 'add new item'

```
<FieldArray name="name_variations" component={renderNameVariations}/>
```

```
const renderNameVariations = ({fields, meta: {touched, error}}) => {
  return (
    <div className="mt5 mb5">
      <label>Name Variations:</label>
      <div className={classNames({'ml30':true, 'form-section':fields.length > 0})}>
        {fields.map((nameVariation, index) =>
          <div key={index} className="ml10">
            <Field name={`${nameVariation}.firstname`} component={renderField} type="text" label="First Name* :" placeholder={'First Name'} />
            <Field name={`${nameVariation}.lastname`} component={renderField} type="text" label="Last Name* :" placeholder={'Last Name'} />
            <Field name={`${nameVariation}.fullname`} component={renderField} type="text" label="Fullname (override)*** :" placeholder={'Fullname'} />
            <div><i>*** If you enter first and last name it will auto populate fullname.  Entering full name will override first and last name.</i></div>
            <button className="btn btn-danger mt5" type="button" title="Remove" onClick={() => fields.remove(index)}><i className="fa fa-trash"></i></button>
          </div>
        )}
      </div>
      <button className="btn btn-secondary m5" type="button" onClick={() => fields.push({})}>Click to Name Variations</button>
    </div>
)};
```

I tried to do the following - inside the renderNameVariations, but I get a React error

```
if(fields && fields.length === 0) {
    fields.push({});
  }
```

Error:  warning.js:36 Warning: setState(...): Cannot update during an existing state transition (such as within `render` or another component's constructor). Render methods should be a pure function of props and state; constructor side-effects are an anti-pattern, but can be moved to `componentWillMount`.

**Comment by bmv437** [comment]
What about setting the initialValues of the form to start with an empty item inside the FieldArray?
```javascript
  <MyForm initialValues={{name_variations: [ { } ] }} />
```

**Comment by Marinolinderhof** [comment]
@bmv437 and @sgentile 
I trying to achieve the same. But then for nested fieldArrays. Like the example.

http://redux-form.com/6.1.1/examples/fieldArrays/

So how can i make sure there is always 1 hobby textfield when you add a member?

I can do that by making the field array a Class component and then on mount 

```
if(fields && fields.length === 0) {
    fields.push({});
  }
```
But that feels expensive. 

![screen shot 2016-10-31 at 14 16 54](https://cloud.githubusercontent.com/assets/3657762/19855596/bcc9b3d2-9f75-11e6-88e7-854b1d9b8312.png)

**Comment by bmv437** [comment]
@Marinolinderhof When you click the `Add Member` button in `renderMembers()`, Instead of `fields.push({})` with an empty object, you would do
```javascript
fields.push({
  "hobbies": [
    null
  ]
});
```
This will set the initial values for the member's hobbies fieldArray, instead of having the `renderHobbies()` function detect whether it needs to pre-populate itself.

**Comment by jay1302** [comment]
how to make 1 member default without having to click on add member?

**Comment by bmv437** [comment]
@jay1302 

set the `initialValues` of the form equal to 
```javascript
{
  "members": [
    {}
  ]
}
```
This will pre-fill the members array with a blank object

**Comment by Marinolinderhof** [comment]
@bmv437 many thanks that works!
for all the immutable users out there user
```
 initialValues: fromJS({ members: [{ hobbies: [null] }] }),
```

or when push
```
() => fields.push(fromJS({ hobies:[null] }))
```

**Comment by helberth** [comment]
I have a fieldarray that depends of a dropdown, in this case I need change the initialValues dynamically every time that dropdown is selected. Any ideas? Thanks.

**Comment by bmv437** [comment]
Hey @helberth, I'm not quite clear on your specific use case, or why it's necessary to do it that way. Could you provide an example?

**Comment by jay1302** [comment]
@bmv437 Thank you!

**Comment by L5eoneill** [comment]
```
componentDidMount() {
  const {my_array_field} = this.props;
    if (my_array_field.length === 0) {
      my_array_field.addField();
  }
}
```

That works in 5.2.5, at least

**Comment by artLopez** [comment]
@bmv437 For setting a form with a default field,  is the `initialValues` data({ "members": [{}] })  set within the reduxForm({}) arguments?

**Comment by bmv437** [comment]
@artLopez It can be set like this:

```javascript
reduxForm({
  form: 'myForm',
  initialValues: {
    "members": [
      {}
    ]
  }
})(MyForm);
```

or like this


```javascript
<MyForm initialValues={{"members": [{}]}} />
```

Either one works, just depends on your preference.

**Comment by artLopez** [comment]
@bmv437 Thank you! I tried to do the initialValues within the reduxForm but it doesn't render the component with a default length of one entry. Would you mind taking a look at my component and seeing if I am doing something wrong? Thanks in advance! 

```
const renderConstraints = ({fields, meta: {touched, error} }) => {
   return(
            <div>
                {fields.map((field, index) =>
                    <div key={index}>
                        <Field name={`${field}.constraintName`}
                               component={RenderTextField} label="Constraint Name" />
                        <Field multiLine={true} rows={2} rowsMax={4}
                               name={`${field}.constraintDescription`} component={RenderTextField} label="Description" />
                        <Field items={category} name={`${field}.constraintType`}
                               component={DropdownField} label="Constraint Category" />
                    </div>
                )}
                <FlatButton onClick={() => fields.push()}>Add Constraint</FlatButton>
            </div>
        )
};

class Constraint extends Component {
    constructor(props, context){
        super(props, context);
    }

    render(){
        return(
            <div>
                <FieldArray name="constraints" component={RenderConstraints} />
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {};
}

Constraint  = reduxForm({
    form: 'myForm',
    initialValues: {'constraints': [{}]},
    destroyOnUnmount: false
})(Constraint);

export default connect(mapStateToProps,null, null, {withRef: true})(Constraint)

**Comment by bmv437** [comment]
Hey @artLopez 

I whipped up a little [jsFiddle](https://jsfiddle.net/bmv437/75rh036o/) to show a working example. when the page loads, it starts off with an empty member as expected.

I looked over you code and couldn't see anything wrong it with it, but then again I can't run your exact code because of some of your custom components. I would suggest a process of elimination, or maybe starting off with a super simple example and working up from there until something breaks. Good luck! I'm curious to see what the issue ends up being.

**Comment by artLopez** [comment]
Hey @bmv437 ,

I apologize for the delay, I was working on some other things. However, I found out was going wrong. I am working on something similar to the [ Wizard Form Example](http://redux-form.com/6.3.1/examples/wizard/) except I am adding the forms dynamically. Thus, I believe it's having a conflict of having multiple components with the same form name. I changed the form name of one component and it was able to have a default field.
